### PR TITLE
Improve some completions on generic object literals

### LINKED ISSUE #34825: autocompletion failing in 3.7.1-rc
**TypeScript Version:**  3.7.1-rc

**Search Terms:**

**Code**

```ts

export type GeMethodsForType<T, G extends string> = { [K in keyof T]:
    T[K] extends () => any ? { name: K, group: G, } : T[K] extends (s: infer U) => any ? { name: K, group: G, payload: U } : never }[keyof T];


class Sample {
    count = 0;
    books: { name: string, year: number }[] = []
    increment() {
        this.count++
        this.count++
    }

    addBook(book: Sample["books"][0]) {
        this.books.push(book)
    }
}
export declare function testIt<T, G extends string>(): (input: any, method: GeMethodsForType<T, G>) => any


const t = testIt<Sample, "Sample">()

const i = t(null, { name: "addBook", group: "Sample", payload: { } }) // auto completion for payload

```

**Expected behavior:**
Propeties of payload should come in autocompletion
**Actual behavior:**
No autocompletion
**Playground Link:** [play ground nightly](https://www.typescriptlang.org/play/index.html?ts=Nightly#code/LAKApgHgDg9gTgFwAQIJ5TEg4mAsmBACxgBMBnAMXgBV0wAeagGmyUgTADtykyE4AlpwDmAPiQBeJAG8kAbQDSSIUgDWYVDABmSagF0AXKCQndivWwgduZJAAoAlJPEBDTqiQB+GUk4uAtmAGSAoswnAwAK5QwVgsAL5IwdTmltY8dmTBQlpgcEgAqk4Sru5ePn6BwaFI4VEx2CxQLqgANjAuJMEFSInBnGAAbnm9cuqaOvoA3KCzIADGrS5ktgDKAVCtmNLGpvNRnMhSAAwzIKZIAEYwMKpZFQFBvPxCwiyoYC5w-ZH+lyPxOQWKRA3YmITzOBgQKHRwyMEXFCEARkAB0+0ihwA1FiERciCj0QcEDiEfE5hdOiQAEI3VR2a63YLrfybMByABEjLuHL0cmOeicO3OiJMBLR3LRUEiZEIDLpDjJoHJ4Gg8GQJDAiy+mC0mPmCAEME4KDAfAAkghGCwsGkuDw+IIRKJHME7EJpQhgm5UCxAkRSLE8ARiOQqHBaBhrdhRMVSqg5qB9pw+ChJKaLVaWWyWBzs1sOS7FSAk8bUwJ0wg7JxIq1WixZJUnhyqbTbhywhFosE8xsC00Wu1OsFZIl4sWgA)

Working fine in 3.7 beta and 3.6.x

[play ground 3.7 beta](https://www.typescriptlang.org/play/index.html?ts=3.7-Beta#code/LAKApgHgDg9gTgFwAQIJ5TEg4mAsmBACxgBMBnAMXgBV0wAeagGmyUgTADtykyE4AlpwDmAPiQBeJAG8kAbQDSSIUgDWYVDABmSagF0AXKCQndivWwgduZJAAoAlJPEBDTqiQB+GUk4uAtmAGSAoswnAwAK5QwVgsAL5IwdTmltY8dmTBQlpgcEgAqk4Sru5ePn6BwaFI4VEx2CxQLqgANjAuJMEFSInBnGAAbnm9cuqaOvoA3KCzIADGrS5ktgDKAVCtmNLGpvNRnMhSAAwzIKZIAEYwMKpZFQFBvPxCwiyoYC5w-ZH+lyPxOQWKRA3YmITzOBgQKHRwyMEXFCEARkAB0+0ihwA1FiERciCj0QcEDiEfE5hdOiQAEI3VR2a63YLrfybMByABEjLuHL0cmOeicO3OiJMBLR3LRUEiZEIDLpDjJoHJ4Gg8GQJDAiy+mC0mPmCAEME4KDAfAAkghGCwsGkuDw+IIRKJHME7EJpQhgm5UCxAkRSLE8ARiOQqHBaBhrdhRMVSqg5qB9pw+ChJKaLVaWWyWBzs1sOS7FSAk8bUwJ0wg7JxIq1WixZJUnhyqbTbhywhFosE8xsC00Wu1OsFZIl4sWgA)

Looks like its getting confused when some properties doesn't contain `payload` property in this case `increment` method , if i change GeMethodsForType to 

```ts
export type GeMethodsForType<T, G extends string> = { [K in keyof T]:
    T[K] extends () => any ? { name: K, group: G, payload?:undefined} : T[K] extends (s: infer U) => any ? { name: K, group: G, payload: U } : never }[keyof T]; // payload added  empty arg methods
```
now autocompletion working fine in 3.7.1-rc  :s

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes some fallout from #33937. Using the intersection was the wrong approach as it led to some nasty effects with conditional and indexed access types:

Screen Shot 2019-10-31 at 10 32 43 AM

This is still not a perfect solution, as it offers too many completions in some cases, like this:

image

I’m planning to iterate on this and hopefully get rid of the invalid ones without un-fixing #30507, but I believe @DanielRosenwasser wants to look at this for 3.7.2.

Fixes #34825

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
How do I run this in the playground @Orta?

**Comment by weswigham** [comment]
You can't - the trigger for the playground builds is borked - Github's broke some experimental API that was in use,

**Comment by orta** [comment]
I've [triggered it manually](https://github.com/orta/make-monaco-builds/commit/ecf882564008b75984e7fed07103eb4d59dfcecc/checks?check_suite_id=297570163): url should be at the end of the build logs

File: tests/cases/fourslash/completionsConditionalMember.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/// <reference path="fourslash.ts" />
+
+////declare function f<T extends string>(
+////  p: { a: T extends 'foo' ? { x: string } : { y: string } }
+////): void;
+////
+////f<'foo'>({ a: { /*1*/ } });
+////f<string>({ a: { /*2*/ } });
```
Comment (sandersn) [inline tests/cases/fourslash/completionsConditionalMember.ts]: Why no tests without type arguments? The new code seems to apply exclusively in this case.

Comment (andrewbranch) [reply tests/cases/fourslash/completionsConditionalMember.ts]: [completionsGenericIndexedAccess2](https://github.com/microsoft/TypeScript/pull/34855/files#diff-10da2e23de6b7208e06ed04e2e1770c5) is without type arguments, which tests the primary way that #33937 was a regression. Almost all the existing tests from #33937 / #32100 are without type arguments, but they weren’t complex enough to fail. This test was to ensure that the new clever stuff _doesn’t_ apply when there _is_ an explicit type argument.

**Pull request merged**